Tab bar improvements

emem365 opened the pull request:
## Description

This PR adds the ability to add padding to user defined indicators in TabBar. Earlier, the TabBar used to ignore indicatorPadding for user-defined indicators. 

To be specific, what I have done is, take this indicatorPadding property and pass it down to the _IndicatorPainter, this way, I apply the padding to the Rect inside which the CustomPainter is going to paint the indicator.

Also, UnderlineTabIndicator, takes in a `insets` parameter, which is set to `indicatorPadding`, and this is really where the indicatorPadding was originally being used. This parameter used to ignore `EdgeInsets.top` and `EdgeInsets.bottom` from the indicatorPadding. *Since I added indicatorPadding to the _IndicatorPainter class, it basically meant that the padding would be applied twice if it is still passed to UnderlineTabIndicator. Therefore, for the time-being, I have commented the line that passes the parameter, while creating the widget(Line-830)(If this feels like a good solution, we can just remove the line). I would really like a maintainers outlook on whether or not I should do this*, and figuring out what else we could do if not this.


## Related Issues

* #59728 

## Tests

Added tests for indicatorPadding being checked for both LTR and RTL text directions. Also added both the test using directional padding, ie, using STEB instead of LTRB.

In these tests, I give the indicator property of TabBar, which takes a Decoration object, a box decoration with a color, as this would cover the entire space available for the indicator. I set some values for the indicatorPadding and then expected to find a a rect being painted with the given padding applied.

## Screenshot
[<img src="https://user-images.githubusercontent.com/47733983/90167029-c5c58a80-ddb8-11ea-9bb2-4310c93ba5fa.jpeg" width=250>](https://user-images.githubusercontent.com/47733983/90167029-c5c58a80-ddb8-11ea-9bb2-4310c93ba5fa.jpeg)
To reproduce the behaviour, copy the contents of [this gist](https://gist.github.com/emem365/9697eecf5c17d9cdadf203178107c026) inside main.dart of a new flutter project.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.
   
<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

emem365 committed:
indicatorPadding now does not get ignored

emem365 committed:
Removing unnecessary comments

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -389,7 +391,10 @@ class _IndicatorPainter extends CustomPainter {
+    final EdgeInsets insets = indicatorPadding.resolve(_currentTextDirection);
+    final Rect rect = Rect.fromLTWH(tabLeft, 0.0, tabRight - tabLeft, tabBarSize.height);
+
+    return insets.deflateRect(rect);
```
This need to handle where insets > original rect, it may become native rect

emem365 replied:
> This need to handle where insets > original rect, it may become native rect
A few assert statements should fix this, i guess. I'll get to it.
>  i think the correct solution is not to clip even without the padding.
I was looking for a way to be able to do that, but I think the way this works is that TabBar draws a Canvas according to the space available to it, and then the indicator is being painted on that.(This is what my understanding of the code says, I might be wrong here.) Painting outside the canvas is obviously a no-no. Could there be a way to be able to increase the size of canvas? That seems like an odd way to solve this because, I think its hard to ever know how big it should be. What do you think about this? Please let me know.

chunhtai replied:
throw a flutter error makes more sense because this is a user define value.

chunhtai reviewed (commented):
Although it is nice to be able to apply padding to the custom indicator, but i think the correct solution is not to clip even without the padding. Does it make sense to somehow allow over painting the shadow outside of the tab bar? cc @HansMuller or @shihaohong ?

emem365 commented:
@chunhtai 

>  i think the correct solution is not to clip even without the padding.

I was looking for a way to be able to do that, but I think the way this works is that TabBar draws a Canvas according to the space available to it, and then the indicator is being painted on that.(This is what my understanding of the code says, I might be wrong here.) Painting outside the canvas is obviously a no-no. Could there be a way to be able to increase the size of canvas? That seems like an odd way to solve this because, I think its hard to ever know how big it should be. What do you think about this? Please let me know.

shihaohong commented:
@chunhtai I suggested increasing the size of the padding to prevent clipping of the shadows in the original issue itself -- Would applying a shadow outside of the boundaries be a valid approach? I was under the assumption that the clipping is expected because the shadow itself "did not fit" into the provided rect, and painting outside of the boundaries of the rect goes against Flutter's layout system since we should not expect any painting to happen outside the rect itself.

chunhtai commented:
> @chunhtai I suggested increasing the size of the padding to prevent clipping of the shadows in the original issue itself -- Would applying a shadow outside of the boundaries be a valid approach? I was under the assumption that the clipping is expected because the shadow itself "did not fit" into the provided rect, and painting outside of the boundaries of the rect goes against Flutter's layout system since we should not expect any painting to happen outside the rect itself.

The other way will be make the shadow part of size calculation, but that maybe a big breaking change, and may be unwanted in different use cases.

That being said, I think this pr is good just by the fact that we can now pad custom indicator.

chunhtai commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -826,7 +827,7 @@ class _TabBarState extends State<TabBar> {
 
     return UnderlineTabIndicator(
-      insets: widget.indicatorPadding,
+//      insets: widget.indicatorPadding,
```
remove this line?

emem365 replied:
@chunhtai Throwing a FlutterError if `(rect.size >= insets.collapsedSize)` isnt true.

emem365 committed:
Throw FlutterError on indicatorPadding

emem365 commented:
@chunhtai Throwing a FlutterError if `(rect.size >= insets.collapsedSize)` isnt true. The way size objects implement >=. This condition holds true only if both height and width of rect is bigger than insets.collapsedSize, which is the ideal confdition for us. So throwing an error if it doesnt satisfy that condition, i.e., if either of the properties is smaller,  for rect made sense.
Also, do you think this PR needs some kind of tests, because I havent been able to understand what I could test with this PR.

chunhtai commented:
@emem365  Sorry for the delay.
In that case, you may need to check both width and height separately.

Yes this need a test. Can you try to see if you can verify the indicatorRect when the indicatorPadding is applied?

emem365 commented:
Ok sure. I'll try to work on this asap

emem365 committed:
Tests for indicatorPadding

emem365 commented:
@chunhtai 
Added tests for indicatorPadding being checked for both LTR and RTL text directions. Also added both the test using directional padding, ie, using STEB instead of LTRB. 

In these tests, I give the indicator property of TabBar, which takes a Decoration object, a box decoration with a color, as this would cover the entire space available for the indicator. I set some values for the indicatorPadding and then expected to find a a rect being painted with the given padding applied.

Similar tests already existed for indicatorPadding being tested against UnderlineIndicator, ie, the default indicator.
Would we want to change the description for these tests that already existed to indicate more clearly that it tests for default indicator or is it fine as is?

chunhtai commented:
@emem365 For some reason I keep thinking I have reviewed this PR, sorry for the delay again.

Instead of changing the existing test description, can you make the new test name more descriptive and indicate that it is testing for custom indicator?

chunhtai commented:
There seems to be some conflict and you may need to rebase off latest master to fix the test failures. Otherwise, this looks good to me except for the test description.

emem365 commented:
> @emem365 For some reason I keep thinking I have reviewed this PR, sorry for the delay again.

@chunhtai  No worries. I understand it must be quite overwhelming with so many issues and PRs

> Instead of changing the existing test description, can you make the new test name more descriptive and indicate that it is testing for custom indicator?

Sure, that makes sense. I'll rebase and change the names soon

emem365 committed:
Migrate test to NNBD

emem365 committed:
Tests Renamed

emem365 commented:
Hey @chunhtai,
I have rebased the branch and renamed the tests. 
Have a look and let me know if anything else needs to be changed.

chunhtai reviewed (approved):
LGTM

emem365 commented:
Thanks 🙌
